cli/src/pcluster/templates/budget_builder.py

```diff
@@ -0,0 +1,79 @@
+from aws_cdk import aws_budgets as budgets
+from aws_cdk.core import Construct
+
+from pcluster.config.cluster_config import BaseClusterConfig
+
+
+class CostBudgets:
+    """Create the budgets based on the configuration file."""
+
+    def __init__(self, scope: Construct, cluster_config: BaseClusterConfig):
+        self.cluster_config = cluster_config
+        self.scope = scope
+        self._create_budgets()
+
+    def _create_budgets(self):
+        """Create the Cfn templates for the list of budgets."""
+        budget_list = self.cluster_config.dev_settings.budgets
+        cfn_budget_list = []
+
+        for index, budget in enumerate(budget_list):
+            cfn_budget = self._add_parameters(budget, index)
+            cfn_budget_list.append(cfn_budget)
+
+        return cfn_budget_list
+
+    def _add_parameters(self, budget, index):
+        """Add each budget's parameters."""
+        budget_data = budgets.CfnBudget.BudgetDataProperty(
+            budget_name=f"CfnBudget{str(index)}{self.cluster_config.cluster_name}{budget.budget_category}",
+            budget_type="COST",
+            time_unit=budget.time_unit,
+            budget_limit=budgets.CfnBudget.SpendProperty(
+                amount=budget.budget_limit.amount,
+                unit=budget.budget_limit.unit,
+            ),
+            cost_filters=(
+                budget.cost_filters
+                if budget.budget_category == "custom"
+                else (
+                    {"TagKeyValue": [f"user:parallelcluster:cluster-name${self.cluster_config.cluster_name}"]}
+                    if budget.budget_category == "cluster"
+                    else ({"TagKeyValue": [f"user:parallelcluster:queue-name${budget.queue_name}"]})
+                )
+            ),
+        )
+        notifications_with_subscribers = add_budget_notifications(budget)
+        budget_id = f"CfnBudget{str(index)}{self.cluster_config.cluster_name}{budget.budget_category}"
+        cfn_budget = budgets.CfnBudget(
+            self.scope, budget_id, budget=budget_data, notifications_with_subscribers=notifications_with_subscribers
+        )
+        return cfn_budget
+
+
+def add_budget_notifications(budget):
```

Comment:
This function is returning either None or a list. It is in general better to have the function return single type of data.
Instead you can move the if condition out of the function:
```
notifications_with_subscribers = add_budget_notifications(budget) if budget.notifications_with_subscribers is not None else None
```
Then in the function you ca remove if

Reply:
Updated!